[ADD] barcode: GS1 nomenclature

Felicious opened the pull request:
## Summary
Add a new document explaining GS1 barcodes in Odoo 15+. Details:
- How to enable and use GS1 barcodes
- 2 workflows: scan lots and quantities in kg
- add new barcode nomenclature to Odoo's default list
- common barcode issues FAQ
- a list of Odoo's default GS1 barcode nomenclature

[Task](https://www.odoo.com/web#model=project.task&id=3292038&active_id=3292038&cids=3) 
[Reference](https://www.odoo.com/slides/slide/answers-to-all-the-faq-about-nomenclature-introduction-2078?fullscreen=1)

## Second edit
- Added a new column in the GS1 nomenclature table for the GS1 field in Odoo. 
- Moved the "create new GS1 rules" section towards the bottom
- Added veryyy detailed instructions on how to use Odoo to identify GS1 barcodes for packagings and quantities. 
- Added a barcodes troubleshooting section based on the GS1 barcodes FAQ you shared in the Inventory smartclass module.

# Third edit
- Confirmed barcode workflows to show in the doc
- Add lots and serial numbers flow
- Shorten the kg flow
- Reduce technical details about regular expressions, check digit in GS1 barcodes

## Reviewers:

In the reviews, I'd like to invite Inventory and Barcode experts to give feedback on:
1. Accuracy of terminology (e.g. A.I., barcode patterns, intro paragraphs) when describing barcodes and their functions to readers who may not have a technical background,
2. More context for when GS1 barcodes are particularly advantageous,
3. Which warehouse workflows would benefit from barcode usage.
4. Pointers on any existing Odoo documents or videos to reference in my learning (:

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/documentation/pull/4352).

odoo-shke commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,267 @@
+
+   Select the :guilabel:`Pencil icon` to view the details of the scan.
+
+   .. image:: gs1_nomenclature/open-pencil-icon.png
```
Can we add a image with the Batch Number set as well on the product?
```suggestion
   .. image:: gs1_nomenclature/open-pencil-icon.png
```

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,267 @@
+
+      odoo.__DEBUG__.services['web.core'].bus.trigger(
+         'barcode_scanned',
+         "0133778856214558300000000110BATCH00000015",
```
Add check in Lots/Serial Numbers page in Inventory.

odoo-shke reviewed (commented):
Off to an awesome start!

Felicious commented:
Hi @odoo/us-doc-review ! This barcode doc was previously reviewed by CRU from Belgium's BSA team, so I think it's good to go for the first round of peer reviews. This is my first time writing a new doc! So I'm open to any feedback on whether I could expound on some concepts better, or if I went too far. (This is also quite a long doc :sweat_smile: ) Because of this, I also didn't go into as much detail as I would usually when explaining the workflows from start to finish. Let me know what you think about dropping those details!

Thanks so much (:

meng-odoo reviewed (changes requested):
Great job on this doc @Felicious! It's super well written and well organized. I just had small typo fixes, wording suggestions, and suggestions for info to add. Please let me know if you have any questions!

Felicious commented:
Thank you so much for reviewing this longgg doc in such detail, @meng-odoo :heart_eyes: 
You even caught punctuation errors in the nomenclature chart at the very end. I appreciate your sharp-eye for mistakes and meticulous-ness.

I implemented your suggestions, and it's ready for another review! Thanks so much :smile:

meng-odoo commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+.. seealso::
+   :ref:`Simplify vendor unit conversions with UoMs<inventory/products/uom-example>`
+
+After the order is placed, navigate to the :menuselection:`Barcode` app to receive the :ref:`vendor
```
The link is to "Receipts" so I think it makes sense for "receive" to be included in the link name 
```suggestion
After the order is placed, navigate to the :menuselection:`Barcode` app to :ref:`receive the vendor
```

meng-odoo commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+
+   Note: the :abbr:`A.I. (application identifier)` for kilograms, `310` + `1`, was used to represent
+   `52.1` kg as a barcode: `000521`. This is because the `1` represents how many digits from the
+   right to place the decimal.
```
```suggestion
   right to place the decimal point.
```

meng-odoo commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+Pattern` is how the sequence of letters or numbers is recognized by the system to contain
+information about the product.
+
+After filling the information, click the :guilabel:`Save & New` button to make another rule or click
```
```suggestion
After filling in the information, click the :guilabel:`Save & New` button to make another rule or click
```

meng-odoo commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+Since GS1 barcodes are challenging to work with, here are some checks to try when the barcodes are
+not working as expected:
+
+#. The :guilabel:`Barcode Nomenclature` setting is set as :menuselection:`Default GS1 Nomenclature`. Jump to
```
```suggestion
#. Ensure that the :guilabel:`Barcode Nomenclature` setting is set as :menuselection:`Default GS1 Nomenclature`. Jump to
```

meng-odoo commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
++---------------------------------+-------------+------------------------------+--------------------+--------------------+
+| Net volume: liters (L)          | Quantity    | (315[0-5])(\\d{6})           | Measure            | Qty in L           |
++---------------------------------+-------------+------------------------------+--------------------+--------------------+
+| Net volume: cubic meters (m^3)  | Quantity    | (316[0-5])(\\d{6})           | Measure            | Qty in m^3         |
```
If there's a way to format the exponents in LaTeX (so they look like exponents instead of `m^3`) that would be great!

Felicious replied:
Did it :heart:

meng-odoo reviewed (changes requested):
Hi @Felicious this doc is looking great! I just had a few small fixes/notes for this final review. Please let me know if you have any questions!

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,396 @@
+   :alt: Choose GS1 from dropdown and click the internal link to see the list of GS1 rules.
+
+To view and edit a list of GS1 *rules* and *barcode patterns* Odoo supports by default, click the
+:guilabel:`➡️ (External link)` icon to the right of the :guilabel:`Barcode Nomenclature` selection.
```
I wasn't able to find a more suitable emoji. Do you think the arrow is close enough, or should I just describe it a the "sideways arrow out of a box" icon?

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+
+.. _barcode/create-GS1-barcode:
+
+Create GS1 barcode
```
Good thinking, Melanie!
```suggestion
Use GS1 barcode
```

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+Then, set up the product barcode by navigating to the intended product form in
+:menuselection:`Inventory --> Products --> Products` and select the product. On the product form,
+click :guilabel:`Edit`. Then, in the :guilabel:`General Information` tab, fill in the
+:guilabel:`Barcode` field with the 14-digit Global Trade Item Number (GTIN) of the product.
```
```suggestion
:guilabel:`Barcode` field with the 14-digit `Global Trade Item Number (GTIN) <https://www.gs1.org/standards/get-barcodes>`_, which is a universal and unique identifying number from GS1. 
```

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+:guilabel:`Receipts` button to view list of vendor receptions to process. Receipts generated from
+abbr:`POs (Purchase Orders)` are listed, but new receipt operations can also be created directly
+through the :menuselection:`Barcode` app using the :guilabel:`Create` button. Click on the
+warehouse operation (`WH/IN`) and scan product barcodes and lot numbers. Use the :guilabel:`✏️
```
```suggestion
warehouse operation (`WH/IN`) and scan product barcodes and lot numbers with a barcode scanner. The scanned product is then displayed on the mobile view. Use the :guilabel:`✏️
```

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+abbr:`POs (Purchase Orders)` are listed, but new receipt operations can also be created directly
+through the :menuselection:`Barcode` app using the :guilabel:`Create` button. Click on the
+warehouse operation (`WH/IN`) and scan product barcodes and lot numbers. Use the :guilabel:`✏️
+(pencil)` button to designate the quantities for specific lot numbers.
```
```suggestion
(pencil)` button to designate the quantities for specific lot numbers. Clicking the pencil button opens a window,

In the edit quantities window, use the number pad and :guilabel:`⬅️ (delete arrow)` button to enter the exact quantity of items in the lot.
```

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,398 @@
+==================
+
+To build GS1 barcodes in Odoo, combine multiple pieces of information using the specified barcode
+pattern. The `application identifier <barcode/operations/app-id>`_ (A.I.) serves as the universal
```
fixed!! thanks for testing these links (:

Felicious commented:
Hi @meng-odoo :smile_cat: Thank you for giving this barcode doc another look! I updated the links and added more clarification on lines 98-100, and 147- 150. I was also able to add LaTex formatting to the exponents at the end! Yayy!

meng-odoo commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,406 @@
++----------------------------------------+-------------+------------------------------+--------------------+--------------------+
+| Net volume: liters (L)                 | Quantity    | (315[0-5])(\\d{6})           | Measure            | Qty in L           |
++----------------------------------------+-------------+------------------------------+--------------------+--------------------+
+| Net volume: cubic meters (:math:`m^3`) | Quantity    | (316[0-5])(\\d{6})           | Measure            | Qty in :math:`m^3` |
```
The exponents in LaTex are super cool! They're looking a little out of place though, since the `math` formatting changes the font. But I found this `superscript` formatting that I think works perfectly!
```
m\ :sup:`3`
```
Using the `math` formatting for just the exponent could also work, although the exponent gets written in the math font:
```
m\ :math:`^3`
```

Felicious replied:
Yayyy!! this workss

meng-odoo commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,406 @@
++----------------------------------------+-------------+------------------------------+--------------------+--------------------+
+| Net weight/volume: ounces (oz)         | Quantity    | (357[0-5])(\\d{6})           | Measure            | Qty in oz          |
++----------------------------------------+-------------+------------------------------+--------------------+--------------------+
+| Net volume: cubic feet (:math:`ft^3`)  | Quantity    | (365[0-5])(\\d{6})           | Measure            | Qty in :math:`ft^3`|
```
Please change these to use the :sup: formatting as well :)

meng-odoo reviewed (approved):
Hi @Felicious, this doc looks great! Everything is explained very clearly. I'm approving with a few small fixes to make before moving on to the next step. Let me know if you have any questions! Then please tag Sam for technical review after you've made those changes :) Great job on this!

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,406 @@
+:guilabel:`Units` are all accounted for, finish the reception by clicking the :guilabel:`Validate`
+button.
+
+.. note::
+   THIS SECTION WILL BE DELETED BEFORE PUBLISHING! It's for the reviewers to test these barcodes
+   easily.
+
+   product: (01)12345678901231
+   qty: (30)00000050
+   Lot#1: (10)LOT0001
+   Lot#2: (10)LOT0002
+
+   .. code-block:: javascript
+
+      // First scan: product
+      odoo.__DEBUG__.services['web.core'].bus.trigger(
+         'barcode_scanned',
+         "0112345678901231",
+         $(".o_web_client")[0],
+      )
+      // Second scan: first Lot#, then enter qty with pencil button
+      odoo.__DEBUG__.services['web.core'].bus.trigger(
+         'barcode_scanned',
+         "10LOT0001",
+         $(".o_web_client")[0],
+      )
+      // Third scan: second Lot#, then enter qty.
+      odoo.__DEBUG__.services['web.core'].bus.trigger(
+         'barcode_scanned',
+         "10LOT0002",
+         $(".o_web_client")[0],
+      )
```
```suggestion

```

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,406 @@
+      :align: center
+      :alt: Scan barcode screen for a reception operation in the Barcode app.
+
+.. note::
+   THIS SECTION WILL BE DELETED BEFORE PUBLISHING! It's for the reviewers to test these barcodes
+   easily!
+
+   On product form for peach, in :guilabel:`General Information` tab, the :guilabel:`Barcode` field
+   =  `00614141000012`
+
+   For kg, A.I : 310
+   decimal number: 1
+   value (6 digits): 000521
+
+   .. code-block:: javascript
+
+      odoo.__DEBUG__.services['web.core'].bus.trigger(
+         'barcode_scanned',
+         "01006141410000123101000521",
+         $(".o_web_client")[0],
+      )
```
```suggestion

```

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,406 @@
+ensure the complete reading of the barcode, it is necessary to add the missing barcode to Odoo's
+list.
+
+.. warning::
```
```suggestion
.. important::
```

Felicious commented:
Hi @samueljlieber :smile: This brand new barcode doc is ready for your technical review :dancers:

samueljlieber commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature/open-pencil-icon.png:
This image is very wide at 1353 x 954, please resize or crop to be under 768px (933px max) wide 🙏

samueljlieber commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature/product-barcodes-page.png:
This image is also very wide at 1349 x 337, please resize or crop to be under 768px (933px max) wide 🙏

samueljlieber commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature/scan-barcode-peaches.png:
Can you please double check compression on this image?

samueljlieber commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature/stock-moves-peach.png:
Can you please double check compression on this image too? 🙏

samueljlieber reviewed (changes requested):
Hi @Felicious! Well done with the content you've added in this doc, really well written and I like the layout! I have a handful of technical corrections and suggestions, please see below and tag me once this PR is ready for another look! Let me know if you have any questions 🙂

Felicious commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,358 @@
+:guilabel:`Traceability` heading, check the box for :guilabel:`Lots & Serial Numbers`.
+
+Then, set up the product barcode by navigating to the intended product form in
+:menuselection:`Inventory --> Products --> Products` and selecting the product. On the product form,
+click :guilabel:`Edit`. Then, in the :guilabel:`General Information` tab, fill in the
```
Wow thanks for noticing this!

Felicious commented:
Hi @samueljlieber, thanks for catching the early line break, all the custom anchor inconsistencies, extra blank spaces, and noticing the pics that didn't get resized! Through this, I realized that if I compress images and then resize them, the images would get uncompressed :skull:

Really appreciate your meticulous review :blush: It's ready for your second look!

samueljlieber reviewed (approved):
Nice work @Felicious! This doc looks good to me 👍 @StraubCreative 🙂

StraubCreative commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,356 @@
+
+.. _barcode/operations/create-GS1-barcode:
+
+Use GS1 barcode
```
Suggestions to be more specific:

1.
```suggestion
Use GS1 barcodes in Odoo
```

2.
```suggestion
Use GS1 in Odoo Barcode
```

StraubCreative commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,356 @@
+patterns concisely. Each barcode pattern begins with a required 2-4 digit :abbr:`A.I. (application
+identifier)`, which corresponds to the rule defined in the system's :ref:`barcode nomenclature list
+<barcode/operations/set-up-barcode-nomenclature>`. By including the appropriate :abbr:`A.I.
+(application identifier)` from the list, Odoo can accurately interpret GS1 barcodes. While most
```
capitalize
```suggestion
(Application Identifier)` from the list, Odoo can accurately interpret GS1 barcodes. While most
```

StraubCreative commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,356 @@
+Use GS1 barcode
+===============
+
+To build GS1 barcodes in Odoo, combine multiple pieces of information using the specified barcode
```
> To build GS1 barcodes in Odoo

This implies Odoo will build the barcodes, meaning, we give Odoo some input data, and Odoo will spit out output data in the form of barcodes.

That doesn't appear to be what's happening? Or at least...that's not clear to me in the text.

It seems like there are rules/conventions to follow, and that's fine, however, I don't see clear instructions or a clear conclusion that Odoo builds barcodes in this section...

StraubCreative commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,356 @@
+
+.. _barcode/operations/lot-setup:
+
+Configuration
```
be more specific 😉 
also why is a **Configuration** heading arriving midway through this section (isn't it usually first?)

StraubCreative commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,356 @@
+   detailed information about the package contents.
+
+.. example::
+   To create a barcode for the product, `Fuji Apple`, enter the 14-digit GTIN `12345678901231` in
```
this isn't helpful because this is demonstrates exactly what *not* to do 😓 

GTINs in particular follow very specific formatting rules, yes, however you can't just make them up (even if you follow the rules for how they're made, such as making sure the character count is right)— these numbers represent uniquely recognizable products in the global supply-chain and therefore have unique IDs. Here's [a table, for example,](https://resources.markon.com/product-catalog/apples-fuji) that shows a bunch of *actual* Fuji Apple GTIN/UPC values.

Would expand the example much more in the next iteration...as it reads right now it communicates "you can just make numbers up and put them here on the form"

StraubCreative commented on content/applications/inventory_and_mrp/inventory/barcode/operations/gs1_nomenclature.rst:
```diff
@@ -0,0 +1,356 @@
+      :alt: Enter 14-digit GTIN into the Barcode field on product form.
+
+.. tip::
+   It is also possible to view a list of all products and barcodes. To access this list, go to
```
wording.
```suggestion
   To view a list of *all* products and their corresponding barcodes in the Odoo database, navigate to...
```

StraubCreative reviewed (approved):
Approving to deploy new documentation
Suggestions below for future iteration(s), thanks!

@robodoo r+
